CFC_DataCollector/trip_analysis/cluster_pipeline.py

```diff
@@ -0,0 +1,130 @@
+import similarity
+from get_database import get_section_db
+import featurization
+import sys
+import math
+
+"""
+This file reads the data from the section database, 
+removes noise from the data, and clusters is. 
+
+The parameters and clustering methods can be easily changed, 
+but based on what works the best, the featurization and clustering 
+works as follows. First, the data is read from the database. 
+For featurization, each section is representated as a start point 
+and an end point. Then, the section is put into bins and the lower 
+half of the bins are removed. Then, the data is clustered using 
+k-means. The parameter for k is currently tested in a range based 
+on the number of elements, but I plan to adjust this range. 
+
+As input, this file accepts an user's uuid from the command line. 
+
+Currently, this file is defaulted not to handle ground truth, but 
+for the purpose of tests, a parameter can be changed to collect ground 
+truth from the database and compare the clustering to ground truth. 
+To change this, in main, change ground_truth to True in the call to 
+read_data. 
+"""
+
+#read the data from the database. If ground_truth is true, it will 
+#take it from the 'color' field of each section in the database. 
+def read_data(uuid, ground_truth=False):
+    sectiondb = get_section_db()
+    sections = sectiondb.find({'user_id' : uuid})
+    if sections.count() == 0:
+        raise Exception('no sections found for user ' + str(uuid))
+    data = []
+    colors = []
+    for section in sections:
+        check(section)
+        start = map(float, section['section_start_point']['coordinates'])
+        end = map(float, section['section_end_point']['coordinates'])
+        data.append({'trip_start_location' : start, 'trip_end_location' : end})
+        if ground_truth:
+            colors.append(section['color'])
+
+    if ground_truth:
+        indices = [] * len(set(colors))
+        for n in colors:
+            if n not in indices:
+                    indices.append(n)
+                    
+        for i in range(len(colors)):
+            colors[i] = indices.index(colors[i])
+
+    return data, colors
+
+#checks that each section has the necessary information
+def check(section):
+    a = find_in_dict(section, 'section_start_point', 'coordinates')
```

Comment:
You can just wrap this in the `Trip` wrapper class instead which handles all the `find_in_dict` stuff for you. Did you find trips that didn't have start or end points?!